File: snippets/ts/src/develop/error_handling.ts
import { TerminalError, RetryableError } from "@restatedev/restate-sdk";
import * as restate from "@restatedev/restate-sdk";

const service = restate.service({
  name: "ErrorHandling",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      // <start_terminal>
      throw new TerminalError("Payment declined.", {
        errorCode: 402,
        metadata: {
          reason: "insufficient_funds",
          paymentId: "payment-123",
        },
      });
      // <end_terminal>
    },
    retryable: async (ctx: restate.Context, name: string) => {
      // <start_retryable>
      await ctx.run("call API", async () => {
          const res = await fetch("https://api.example.com/data");
          if (!res.ok) {
              const retryAfter = res.headers.get("Retry-After");
              throw new RetryableError("Rate limited", {
                  retryAfter: { seconds: Number(retryAfter ?? 30) },
              });
          }
          return res.json();
      }, { maxRetryAttempts: 10 });
      // <end_retryable>
    },
    retryableFrom: async (ctx: restate.Context, name: string) => {
      const cause = new Error("something failed");
      // <start_retryable_from>
      throw RetryableError.from(cause, {
          retryAfter: { seconds: 30 },
      });
      // <end_retryable_from>
    },
  },
});

const billingService = restate.service({
  name: "BillingService",
  handlers: {
    charge: async (_ctx: restate.Context, paymentId: string) => {
      throw new TerminalError("Payment declined.", {
        errorCode: 402,
        metadata: { reason: "insufficient_funds", paymentId },
      });
    },
  },
});

// <start_terminal_metadata_caller>
async function checkout(ctx: restate.Context, paymentId: string) {
  try {
    await ctx.serviceClient(billingService).charge(paymentId);
    return { status: "charged" };
  } catch (error) {
    if (error instanceof TerminalError) {
      return {
        status: "declined",
        code: error.code,
        reason: error.metadata?.reason,
      };
    }
    throw error;
  }
}
// <end_terminal_metadata_caller>

// <start_pause_error>
async function chargeCard(ctx: restate.Context, paymentId: string) {
  return ctx.run("charge card", async () => {
    const providerUrl = process.env.PAYMENT_PROVIDER_URL;
    if (providerUrl === undefined) {
      throw new restate.PauseError("Payment provider is not configured");
    }

    const response = await fetch(`${providerUrl}/payments/${paymentId}`, {
      method: "POST",
    });
    if (!response.ok) {
      throw new Error(`Payment provider returned HTTP ${response.status}`);
    }
    return response.json();
  });
}
// <end_pause_error>

// <start_journal_mismatch>
const paymentService = restate.service({
  name: "PaymentService",
  handlers: {
    charge: async (_ctx: restate.Context, paymentId: string) => {
      return `Charged ${paymentId}`;
    },
  },
  options: {
    onJournalMismatchErrors: "pause",
  },
});
// <end_journal_mismatch>

// <start_as_terminal>
class MyValidationError extends Error {}

const greeter = restate.service({
  name: "greeter",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      if (name.length === 0) {
        throw new MyValidationError("Length too short");
      }
      return `Hello ${name}`;
    },
  },
  options: {
    asTerminalError: (err) => {
      if (err instanceof MyValidationError) {
        // My validation error is terminal
        return new restate.TerminalError(err.message, { errorCode: 400 });
      }
    },
  },
});
// <end_as_terminal>

File: snippets/python/src/develop/error_handling.py
from datetime import timedelta

import restate
from restate import Service, Context

my_service = Service("MyService")


@my_service.handler()
async def my_handler(ctx: Context, arg):
    # <start_terminal>
    from restate.exceptions import TerminalError

    raise TerminalError(
        "Payment declined.",
        status_code=402,
        metadata={
            "reason": "insufficient_funds",
            "payment_id": "payment-123",
        },
    )
    # <end_terminal>


from restate.exceptions import TerminalError

billing_service = Service("BillingService")


@billing_service.handler()
async def charge(ctx: Context, payment_id: str):
    raise TerminalError(
        "Payment declined.",
        status_code=402,
        metadata={"reason": "insufficient_funds", "payment_id": payment_id},
    )


# <start_terminal_metadata_caller>
async def checkout(ctx: Context, payment_id: str):
    try:
        await ctx.service_call(charge, arg=payment_id)
        return {"status": "charged"}
    except TerminalError as err:
        return {
            "status": "declined",
            "code": err.status_code,
            "reason": err.metadata.get("reason") if err.metadata else None,
        }


# <end_terminal_metadata_caller>


def undo_transaction():
    pass


@my_service.handler()
async def my_handler_2(ctx: Context, arg):
    # <start_run>
    try:
        # Await a ctx.run_typed raising an error
        def do_transaction():
            raise TerminalError("Can't write")

        await ctx.run_typed("do transaction", do_transaction)
    except TerminalError as err:
        # Handle the terminal error raised by ctx.run_typed
        # For example, undo previous actions...
        await ctx.run_typed("undo transaction", undo_transaction)
        # ...and propagate the error
        raise err
    # <end_run>


def make_request():
    pass


@my_service.handler()
async def my_handler_3(ctx: Context, arg):
    # <start_retryable>
    from datetime import timedelta
    from restate.exceptions import RetryableError

    raise RetryableError(
        "Service temporarily unavailable",
        retry_after=timedelta(seconds=30),
    )
    # <end_retryable>


@my_service.handler()
async def my_handler_4(ctx: Context, arg):
    # <start_retryable_run>
    from datetime import timedelta
    from restate.exceptions import RetryableError

    async def call_external_api():
        response = await make_request()
        if response.status == 429:
            retry_after = int(response.headers.get("Retry-After", "30"))
            raise RetryableError(
                "Rate limited",
                retry_after=timedelta(seconds=retry_after),
            )
        return response.data

    result = await ctx.run_typed(
        "call API",
        call_external_api,
        restate.RunOptions(max_attempts=5),
    )
    # <end_retryable_run>
